### PR TITLE
🛠️ Feat/1 Setup Repo

### PR DESCRIPTION
**Pull request by demariadaniel**
- Creates basic React UI, Express API, and Types package, using Vite for set up
- Set up unit testing framework using Node Test Runner
- Adds basic README docs

### CODE REVIEW

File: apps/api/package.json
Comment (justincorrigible) [inline apps/api/package.json:17]: think you mean this one to be in "engines" instead, no?
not sure `npm i node` is a useful thing, but I'm very curious to see what this does!

Comment (demariadaniel) [reply apps/api/package.json:17]: Yes (I'm curious too haha) but this PR is still in an extremely unfinished state so no need for feedback yet

File: packages/types/src/demoData.mts
Comment (demariadaniel) [inline packages/types/src/demoData.mts:22]: Needed some demo data, so I created this based on early planning sessions. Expect this to completely change based on outcome of https://app.zenhub.com/workspaces/pcgl---pancanadian-genome-library-659eed8bc4fe0c04890cff77/issues/gh/pan-canadian-genome-library/daco/2

File: apps/api/src/main.mts
Comment (ciaranschutte) [inline apps/api/src/main.mts:24]: is this just a typo? should be .mts instead of .mjs?

File: apps/ui/eslint.config.js
Comment (ciaranschutte) [inline apps/ui/eslint.config.js:21]: nothing wrong here, just a heads up - I think this will enforce far stricter hook rules than we're used to. specifically regarding hook dependencies

File: apps/ui/package.json
Comment (ciaranschutte) [inline apps/ui/package.json:8]: I think you can just have tsc only typecheck (not build) and leave the building to Vite - afaik it works on ts files

File: packages/types/vite.config.ts
Comment (ciaranschutte) [inline packages/types/vite.config.ts:12]: curious why rollupOptions need to be defined with just an entry point. I think it's covered by lib already?

File: apps/ui/src/main.tsx
Comment (ciaranschutte) [inline apps/ui/src/main.tsx:6]: what does the `!` do here?

**Comment by ciaranschutte** [comment]
looks good! I'd recommend adding prettier + husky at this point. standard in our stack. 
husky for pre commit checks and prettier for consistent styling
prettier-organise-imports package is useful too

File: apps/api/src/main.mts
Comment (ciaranschutte) [inline apps/api/src/main.mts:29]: at this setup point, likely good to add .env config - usually using the dotenv library - things like PORT should be configurable based on env.

**Comment by ciaranschutte** [comment]
license could be added also at this point. kinda type that usually gets forgotten later!

File: package.json
Comment (ciaranschutte) [inline package.json:21]: extra package.json?

Comment (joneubank) [reply package.json:21]: pnpm workspace root package, normal to have

File: apps/api/src/main.mts
Comment (joneubank) [reply apps/api/src/main.mts:29]: Good idea for a next PR is adding template for .env config.

File: apps/ui/src/main.tsx
Comment (joneubank) [reply apps/ui/src/main.tsx:6]: insists this will be not null. its a shortcut around TS type checking for either "I know better" or "Its ok for things to break here". This is becoming a pretty standard setup for React apps since if this node is missing the whole website will fail to render and so there isn't really a fallback.

File: apps/ui/eslint.config.js
Comment (joneubank) [reply apps/ui/eslint.config.js:21]: This seems like a good thing to start with, no?

File: apps/api/src/main.mts
Comment (joneubank) [reply apps/api/src/main.mts:24]: TS with EMS imports require .js file extension.

File: apps/api/src/main.mts
Comment (joneubank) [inline apps/api/src/main.mts:23]: Whats this about? I wouldn't know how to action this todo.

File: apps/api/src/main.mts
Comment (joneubank) [inline apps/api/src/main.mts:39]: Try `_req`. The underscore prefix indicates this is a placeholder parameter and is unused. TS will be happy with that. This TODO is why it may not make sense to enforce no unused params/locals.

File: apps/api/src/main.mts
Comment (joneubank) [inline apps/api/src/main.mts:47]: Does `app.listen` create a new object, or could you just export `app` where it is declared and get the same effect? I think it would be clearer what is happening if you export the const where it is declared.

File: apps/api/src/tests/index.mjs
Comment (joneubank) [inline apps/api/src/tests/index.mjs:21]: Thank you for the `node:` prefix on node api imports.

File: apps/api/src/tests/index.mjs
Comment (joneubank) [inline apps/api/src/tests/index.mjs:56]: Nice example usage. I think we can remove this from our setup though :)

File: apps/api/src/tests/index.mjs
Comment (joneubank) [inline apps/api/src/tests/index.mjs:68]: Again, nice examples. Let's remove them from our setup.

File: apps/ui/README.md
Comment (joneubank) [inline apps/ui/README.md:1]: Lets replace this template readme with one related to the application we are actually providing. 

For inspiration you can look at the OHCRN repo for package readme examples:
https://github.com/OHCRN/platform/blob/develop/apps/consent-api/README.md
https://github.com/OHCRN/platform/blob/develop/apps/consent-ui/README.md

File: apps/ui/index.html
Comment (joneubank) [inline apps/ui/index.html:5]: nice!

File: packages/types/src/types.ts
Comment (joneubank) [inline packages/types/src/types.ts:77]: Type tagging? Looks like we're applying this inconsistently (used on revisions but not Project). Curious what the motivation for this is here.

**Review by joneubank — CHANGES_REQUESTED** [review]
Really great starting point for this project, thanks. Left a handful of comments. In addition I want to suggest:

- Readmes:
   - no readme should say this is an Overture project. Its great to use some templating from other projects but we need to make sure the READMEs are all relevant to our project
   - Important to have a readme in each of the apps/packages, and we want these to be relevant from day 1. This means a relevant title, not a template description, and meaningful sections with configuration infomration, dev instructions, and any unique details a user or developer may be interested in
- TSConfig:
   - can you add some documentation explaining the split in tsconfigs between .app and .node? Or if this isnt needed, lets simplify the setup.
   - since we don't have scripts using one tsconfig vs the other, I'm pretty sure we only need a single tsconfig per project
   - It would be good to have a global tsconfig that all the projects reference. This would enforce standard linting rules (strict, unsused parameters, switch fallthrough, unchecked index, etc.) . Each package would then extend this base config and add their custom settings for bundler and target etc.
- Testing:
   - love seeing the node test demos, but we don't need these committed and left in our repo
- Prettier - Can we setup a global prettier config right from the start?

File: apps/ui/eslint.config.js
Comment (ciaranschutte) [reply apps/ui/eslint.config.js:21]: @joneubank 100%! as I say, nothing wrong at all, just a heads up. We had to tone down our eslint usage in a previous project (can't recall exactly which one) because we had so many things to fix. 

I agree - very good thing to start with - nice one @demariadaniel !

File: README.md
Comment (demariadaniel) [reply README.md]: Re: Overture, I'm removing templated references, I was waiting for guidance on copy re: project summary and will review past meetings w/ Alexis

File: apps/ui/README.md
Comment (demariadaniel) [reply apps/ui/README.md:1]: Thanks, I need permissions to view

File: packages/types/src/types.ts
Comment (demariadaniel) [reply packages/types/src/types.ts]: Removed, this was more for noting how we should define our data as types, which is now implemented in the demo data

File: apps/api/src/main.mts
Comment (demariadaniel) [reply apps/api/src/main.mts:23]: Added specific comment. The goal is to get our export from the main package export, not from a specific file in `src`. So the `/packages` build needs some more configuring.

File: packages/types/package.json
Comment (demariadaniel) [reply packages/types/package.json:2]: I've updated `types` to `data-model` following the example in Lyric: https://github.com/overture-stack/lyric/tree/main/packages
This better expresses what the package is used for but naming is hard so I'm open to other suggestions

File: README.md
Comment (demariadaniel) [reply README.md]: Updated

File: .gitignore
Comment (demariadaniel) [reply .gitignore]: Updated

File: apps/ui/tsconfig.app.json
Comment (demariadaniel) [reply apps/ui/tsconfig.app.json:17]: Added

File: apps/api/.gitignore
Comment (demariadaniel) [reply apps/api/.gitignore:1]: I was able to move everything to the root; the project level gitignores were all duplicates. I deleted the files; they can be added as needed.

File: docs/1-Overview.md
Comment (demariadaniel) [reply docs/1-Overview.md:1]: All references removed

File: apps/api/src/main.mts
Comment (demariadaniel) [reply apps/api/src/main.mts:29]: Are you guys OK if I create a new issue to cover:
- Add env support
- Add Husky
- Add Prettier
- Add License
then add those in a follow up PR?

File: apps/api/src/main.mts
Comment (demariadaniel) [reply apps/api/src/main.mts:47]: `.listen` does return a `server` which is different from `app` (returned from `express()`)
So I wouldn't be able to export `app` and call `.close`

See:
- https://expressjs.com/en/advanced/healthcheck-graceful-shutdown.html
- https://expressjs.com/en/5x/api.html#app.listen

File: apps/api/src/tests/index.mjs
Comment (demariadaniel) [reply apps/api/src/tests/index.mjs]: Removed. Kept only the first test case.

File: apps/ui/tsconfig.app.json
Comment (demariadaniel) [reply apps/ui/tsconfig.app.json:20]: Yeah I think linter is appropriate. It's always good to remove unused imports but with variables it can get tricky when there are unused positional arguments. The Express `req` example is the perfect example.

File: apps/api/src/main.mts
Comment (demariadaniel) [reply apps/api/src/main.mts]: Thank you did not know about that trick

File: apps/ui/tsconfig.json
Comment (demariadaniel) [reply apps/ui/tsconfig.json:5]: This is Vite React boilerplate. It automatically sets up the React project this way. I can look into this tomorrow but assumed it was a pre-req to get React up and running. I don't see how the tsconfig.node is used other than for the Vite config. (see `tsconfig.node` -> `include`)

File: apps/ui/tsconfig.node.json
Comment (demariadaniel) [reply apps/ui/tsconfig.node.json]: ES2022 / ES2023 here are Vite template boilerplate. I can't find documentation giving an explicit reason for these vs. ESNext. I assume it's to use a modern but stable version, without introducing experimental features.

File: apps/ui/tsconfig.json
Comment (demariadaniel) [reply apps/ui/tsconfig.json]: Combined into a single file, and the `run dev` script still works.

File: packages/types/vite.config.ts
Comment (demariadaniel) [reply packages/types/vite.config.ts]: Eh there was some back and forth during setup. This package, I think I chose the wrong Vite template to start with and have since removed these options.

File: README.md
Comment (demariadaniel) [reply README.md]: Updated dependencies & setup

File: apps/ui/README.md
Comment (demariadaniel) [reply apps/ui/README.md]: Removed auto-generated Vite content until this is resolved

File: apps/api/src/main.mts
Comment (ciaranschutte) [inline apps/api/src/main.mts:26]: keep imports at top of file

File: apps/api/src/main.mts
Comment (ciaranschutte) [inline apps/api/src/main.mts:36]: setting `cors` middleware at app level means it's enabled for all app handlers
https://expressjs.com/en/resources/middleware/cors.html
I think the first and second example both made it in

File: apps/api/src/tests/index.mjs
Comment (ciaranschutte) [inline apps/api/src/tests/index.mjs:25]: no async operation here?

File: apps/api/vite.config.ts
Comment (ciaranschutte) [inline apps/api/vite.config.ts:23]: vite config for lib. doesn't make sense to me here

File: apps/ui/src/App.tsx
Comment (ciaranschutte) [inline apps/ui/src/App.tsx:40]: this makes me wonder if prettier + husky is working correctly.

File: apps/ui/src/App.tsx
Comment (ciaranschutte) [inline apps/ui/src/App.tsx:24]: good to have consistent code style throughout a project! 
We usually use `const App = () => {}` style but I'm open to this stylistic choice, let's keep it consistent for the components to come too

File: tsconfig.json
Comment (ciaranschutte) [inline tsconfig.json:13]: won't this remap all imports to a subdir of node_modules?

File: apps/api/src/main.mts
Comment (ciaranschutte) [reply apps/api/src/main.mts:42]: why export anything? It's server entry point

**Comment by ciaranschutte** [comment]
A few comments/bits of advice. Overall looking good :thumbs-up:
I'll leave it up to you & others on the project to decide how actionable these are.

- env <br/>
https://github.com/Pan-Canadian-Genome-Library/daco/pull/7/commits/9a14e7c1da573a4019df438feef510532ea0ff2c
.env support added here, but no .env.schema added which is helpful to give developers are starting point.


- no need to export `server`, I found that confusing. `app.listen` is the magic here.

- it's easier to give prettier a specific folder like `src` to modify rather than have a tonne of ignore options. When people add other files, they may have to add to prettier ignore. aim is sensible default behaviour. <br/><br/> I don't understand the `vite.config.js` usage for server.  I don't see it in the Vite boilerplates. I think you may be basing it off the library config I shared with you for iobio. In `iobio` the end goal was a package, a library to install and use in a project. <br/><br/>It's different here.
We don't install the server or the ui as packages, so we don't need to build them as libraries.
The simplest server setup is to install dependencies then run the server (vite). 
I believe no additional config needed... just run vite.


- `app.use(cors())` is setting up middleware cors for all requests on app level, no need to use it later on individual handlers if you set it up like this.

File: apps/api/src/main.mts
Comment (demariadaniel) [reply apps/api/src/main.mts:26]: Moved

File: apps/api/src/main.mts
Comment (demariadaniel) [reply apps/api/src/main.mts]: Ooh sorry I see, yeah it looks like I mixed their examples. Updated.

File: apps/api/src/tests/index.mjs
Comment (demariadaniel) [reply apps/api/src/tests/index.mjs]: Leftover from earlier test definitions, removed

File: apps/ui/src/App.tsx
Comment (demariadaniel) [reply apps/ui/src/App.tsx:40]: Which aspect? 
Running `pnpm run prettier` shows this file as `unchanged`, editor makes no additional changes on save (but will change import order, for example)

**Comment by demariadaniel** [comment]
> * it's easier to give prettier a specific folder like `src` to modify rather than have a tonne of ignore options. When people add other files, they may have to add to prettier ignore. aim is sensible default behaviour. 

Good call, updated npm script to target `./packages` & `./apps`

File: apps/api/vite.config.ts
Comment (demariadaniel) [reply apps/api/vite.config.ts:23]: Yeah I had a feeling this wasn't necessary, let me remove

**Comment by demariadaniel** [comment]
> I don't understand the `vite.config.js` usage for server.  I don't see it in the Vite boilerplates. I think you may be basing it off the library config I shared with you for iobio. In `iobio` the end goal was a package, a library to install and use in a project. It's different here.
>   We don't install the server or the ui as packages, so we don't need to build them as libraries.
>   The simplest server setup is to install dependencies then run the server (vite).
>   I believe no additional config needed... just run vite.

So given we don't need to build a package, I don't know if we need Vite at all for the API. 
I used Vite for setup but it's actually just a Node/Express app.
We could make a `run:prod` script which essentially does the same thing as `run:dev` at this stage.
This is essentially how platform-api is built: https://github.com/icgc-argo/platform-api/blob/develop/package.json#L15

> * right now the dev script is using tsx with an env flag. for prod, looks like vite should be used, so you'll likely have to use something like https://www.npmjs.com/package/dotenv to read env vars.

The flag is a native Node flag: https://nodejs.org/en/learn/command-line/how-to-read-environment-variables-from-nodejs 
I wanted to leverage modern Node features as much as possible to keep it minimal. See above comment, if Vite is not a requirement then is DotEnv needed?

File: tsconfig.json
Comment (demariadaniel) [reply tsconfig.json:13]: Good catch, previously had added this to the API config when I was still getting things running. Isn't needed now. Removed.

File: apps/api/src/main.mts
Comment (demariadaniel) [reply apps/api/src/main.mts]: Triggering the test runner starts the server, so I needed a reference in the test file to close it.
I've removed the export and replaced it with `process.exit()` in the tests.
We can add server-endpoint testing when needed.

File: apps/ui/src/App.tsx
Comment (ciaranschutte) [reply apps/ui/src/App.tsx:40]: I missed that it's a template string with newlines.
just a really odd variable I guess.

**Review by ciaranschutte — APPROVED** [review]
Looks good. Approving.

Not sure why there's a `docs` folder with empty files. 
It's a nice thought but unless it's definitely going to worked it's just adding empty files to a repo.
if keeping , might be worth adding some content even if very minimal for now.

**Comment by demariadaniel** [comment]
> Looks good. Approving.
> 
> Not sure why there's a `docs` folder with empty files. It's a nice thought but unless it's definitely going to worked it's just adding empty files to a repo. if keeping , might be worth adding some content even if very minimal for now.

Thanks @ciaranschutte there's a handful of Readme items to resolve so I created a follow up ticket: https://app.zenhub.com/workspaces/pcgl---pancanadian-genome-library-659eed8bc4fe0c04890cff77/issues/gh/pan-canadian-genome-library/daco/14

These are template files from Overture, will revisit

**Pull request merged**
